architecture/src/androidLibMain/kotlin/observable/Observable.kt

```diff
@@ -0,0 +1,186 @@
+import androidx.lifecycle.Observer
+import androidx.lifecycle.asLiveData
+import com.splendo.kaluga.flow.BaseFlowable
+import kotlin.properties.ObservableProperty
+import kotlin.properties.ReadOnlyProperty
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+actual abstract class Observable<T> : ReadOnlyProperty<Any?, ObservableResult<T>> {
+
+    /**
+     * [LiveData] syncing with this observable
+     */
+    abstract val liveData: LiveData<T>
+
+    /**
+     * Adds an [Observer] to this Observable
+     * @param owner The [LifecycleOwner] of the [Observer]
+     * @param observer The [Observer] to observe this Observable
+     */
+    fun observe(owner: LifecycleOwner, observer: Observer<T>) {
+        liveData.observe(owner, observer)
+    }
+
+    override fun getValue(thisRef: Any?, property: KProperty<*>): ObservableResult<T> {
+        return liveData.value?.let { ObservableResult.Result(it) } ?: ObservableResult.Nothing()
+    }
+}
+
+/**
+ * Simple [Observable] that takes an initial value
+ * @param initialValue The initial value of the Observable
+ */
+class DefaultObservable<T>(initialValue: T) : Observable<T>() {
+    override val liveData = MutableLiveData(initialValue)
+}
+
+/**
+ * [Observable] whose initial value matches a given [ReadOnlyProperty]
+ * @param readOnlyProperty The [ReadOnlyProperty] to match with the initial value
+ */
+class ReadOnlyPropertyObservable<T>(readOnlyProperty: ReadOnlyProperty<Any?, T>) : Observable<T>() {
+    private val value by readOnlyProperty
+    override val liveData: LiveData<T> = MutableLiveData(value)
+}
+
+/**
+ * [Observable] whose value matches a given [Flow]
+ * @param flow The [Flow] whose value to match
+ * @param coroutineScope The [CoroutineScope] on which the observe the [Flow]
+ */
+class FlowObservable<T>(flow: Flow<T>, coroutineScope: CoroutineScope) : Observable<T>() {
+    override val liveData: LiveData<T> = flow.asLiveData(coroutineScope.coroutineContext)
+}
+
+actual abstract class Subject<T>(private val coroutineScope: CoroutineScope) : Observable<T>(), ReadWriteProperty<Any?, ObservableResult<T>> {
+
+    protected abstract val providerLiveData: LiveData<T>
+    private val mediatorLiveData = MediatorLiveData<T>()
+
+    /**
+     * [MutableLiveData] syncing with this Subject
+     */
+    override val liveData: MutableLiveData<T> = mediatorLiveData
+    protected abstract val liveDataObserver: Observer<T>
+
+    protected fun initialize() {
+        mediatorLiveData.addSource(providerLiveData) {
+            value -> mediatorLiveData.postValue(value)
+        }
+        coroutineScope.launch {
+            liveData.observeForever(liveDataObserver)
+            while (isActive) {
+                delay(100)
```

Comment:
Why this delay and why 100?

Reply:
good question actually.

Reply:
Just to limit the calls, but I can remove it

Reply:
Removed